### PR TITLE
Add PP and TP support for XPU

### PR DESCRIPTION
**Pull request by Alokksinha00**
## Motivation

To support Pipeline and Tensor parallelism in XPU
## Accuracy Tests
test/srt/test_pp_single_node.py

## Benchmarking and Profiling
PP-->
python -m sglang.bench_one_batch_server --batch-size 2 --input-len 256 --output-len 32 --model "meta-llama/Llama-3.1-8B-Instruct" --device xpu --host 127.0.0.1 --port 8000 --mem-fraction-static 0.8 --profile --pp-size 2

batch size: 2
input_len: 256
output_len: 32
latency: 5.73 s
ttft: 2.79 s
last generation throughput: 5.89 tok/s
input throughput: 183.22 tok/s
output throughput: 21.82 tok/s

TP-->
python -m sglang.bench_one_batch_server --batch-size 1 --input-len 256 --output-len 32 --model "Qwen/Qwen3-1.7B" --device xpu --host 127.0.0.1 --port 8000 --mem-fraction-static 0.8 --profile --tp-size 2 --disable-radix-cache

batch size: 1
input_len: 256
output_len: 32
latency: 2.17 s
ttft: 0.21 s
last generation throughput: 0.00 tok/s
input throughput: 1200.02 tok/s
output throughput: 16.33 tok/s

python3 -m sglang.bench_one_batch --batch-size 4 --input 32 --output 32 --model meta-llama/Llama-3.2-3B-Instruct --dtype bfloat16 --tp 2 --trust-remote-code --mem-fraction-static 0.5 --device xpu

Prefill. latency: 2.90453 s, throughput:     44.07 token/s
Decode 0. Batch size: 4, latency: 0.09912 s, throughput:     40.36 token/s
Decode 1. Batch size: 4, latency: 0.03683 s, throughput:    108.61 token/s
Decode 2. Batch size: 4, latency: 0.03683 s, throughput:    108.60 token/s
Decode 3. Batch size: 4, latency: 0.03655 s, throughput:    109.45 token/s
Decode 4. Batch size: 4, latency: 0.03650 s, throughput:    109.58 token/s
Decode.  median latency: 0.03644 s, median throughput:    109.77 token/s
Total. latency:  4.099 s, throughput:     62.46 token/s
Benchmark ...
Prefill. latency: 0.06894 s, throughput:   1856.77 token/s
Decode 0. Batch size: 4, latency: 0.03836 s, throughput:    104.28 token/s
Decode 1. Batch size: 4, latency: 0.03801 s, throughput:    105.22 token/s
Decode 2. Batch size: 4, latency: 0.03632 s, throughput:    110.14 token/s
Decode 3. Batch size: 4, latency: 0.03617 s, throughput:    110.59 token/s
Decode 4. Batch size: 4, latency: 0.03615 s, throughput:    110.66 token/s
Decode.  median latency: 0.03630 s, median throughput:    110.19 token/s
Total. latency:  1.199 s, throughput:    213.51 token/s

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by Alokksinha00** [comment]
Can you please review, also help to trigger ci ? @sglang-bot 
@yizhang2077, @mingfeima

File: python/sglang/srt/utils/common.py [hunk 1]
```diff
@@ -1299,34 +1211,37 @@ def point_to_point_pyobj(
     dst: int = 1,
 ):
     """Send data from src to dst in group using DeviceToDevice communication."""
-    device = torch.get_device_module().current_device()
+    # Generic way to get current device for any accelerator ('cuda', 'xpu', 'hpu', etc)
+    device_module = torch.get_device_module()
+    current_device = torch.device(device_module.__name__.split(".")[-1])
```
Comment (polisettyvarma) [inline python/sglang/srt/utils/common.py]: current_device = device_module.current_device()

**Review by polisettyvarma — CHANGES_REQUESTED** [review]

File: python/sglang/srt/distributed/parallel_state.py [hunk 2]
```diff
@@ -303,48 +278,41 @@ def __init__(
...
+        device_id = 0 if IS_ONE_DEVICE_PER_PROCESS else local_rank
         if is_cuda_alike():
-            device_id = (
-                0 if envs.SGLANG_ONE_VISIBLE_DEVICE_PER_PROCESS.get() else local_rank
-            )
             self.device = torch.device(f"cuda:{device_id}")
         elif _is_npu:
-            self.device = torch.device(f"npu:{local_rank}")
+            self.device = torch.device(f"npu:{device_id}")
+        elif is_xpu:
```
Comment (polisettyvarma) [inline python/sglang/srt/distributed/parallel_state.py]: introduce _is_xpu, this is incorrect

**Review by polisettyvarma — CHANGES_REQUESTED** [review]

File: python/sglang/srt/distributed/parallel_state.py (hunk 2 above)
Comment (Alokksinha00) [reply python/sglang/srt/distributed/parallel_state.py]: ok done

File: python/sglang/srt/utils/common.py (hunk 1 above)
Comment (Alokksinha00) [reply python/sglang/srt/utils/common.py]: done

File: python/sglang/srt/utils/common.py [hunk 3]
```diff
@@ -1182,23 +1182,24 @@ def point_to_point_pyobj(
     dst: int = 1,
 ):
     """Send data from src to dst in group using DeviceToDevice communication."""
+    # Generic way to get current device for any accelerator ('cuda', 'xpu', 'hpu', etc)
+    device_module = torch.get_device_module()
+    current_device = device_module.current_device()
```
[the remark this answers is not in the document]
Comment (Alokksinha00) [reply python/sglang/srt/utils/common.py:1332]: done

File: python/sglang/srt/distributed/parallel_state.py [hunk 4]
```diff
@@ -55,12 +55,17 @@
...
     is_shm_available,
     is_xpu,
+    supports_custom_op,
 )
 from sglang.srt.utils.custom_op import register_custom_op
 
 _is_npu = is_npu()
 _is_cpu = is_cpu()
 _is_xpu = is_xpu()
+_supports_custom_op = supports_custom_op()
```
Comment (polisettyvarma) [inline python/sglang/srt/distributed/parallel_state.py:65]: seems not used, why this change ?

File: python/sglang/srt/distributed/parallel_state.py [hunk 5]
```diff
@@ -313,6 +318,17 @@ def __init__(
...
+        device_id = 0 if IS_ONE_DEVICE_PER_PROCESS else local_rank
+        if is_cuda_alike():
+            self.device = torch.device(f"cuda:{device_id}")
+        elif _is_npu:
+            self.device = torch.device(f"npu:{device_id}")
+        elif _is_xpu:
+            self.device = torch.device(f"xpu:{device_id}")
+        else:
+            self.device = torch.device("cpu")
+        self.device_module = torch.get_device_module(self.device)
```
Comment (polisettyvarma) [inline python/sglang/srt/distributed/parallel_state.py:330]: they already exist in line 268 to 281

File: python/sglang/srt/utils/common.py [hunk 6]
```diff
@@ -1336,10 +1340,7 @@ def point_to_point_pyobj(
     if rank == src:
         p2p_works = []
         if len(data) == 0:
-            tensor_size = torch.tensor(
-                [0],
-                dtype=torch.long,
-            )
+            tensor_size = torch.tensor([0], dtype=torch.long, device=current_device)
```
Comment (polisettyvarma) [inline python/sglang/srt/utils/common.py:1343]: moving from cpu to device, what are you trying to solve ?

**Review by polisettyvarma — CHANGES_REQUESTED** [review]
